File: lib/src/parser/page_layout_parser.dart
import 'package:musicxml_parser/src/models/page_layout.dart';
import 'package:musicxml_parser/src/parser/xml_helper.dart';
import 'package:xml/xml.dart';

/// Parser for <page-layout> elements.
class PageLayoutParser {
  /// Parses an [XmlElement] into a [PageLayout] object.
  PageLayout parse(XmlElement element) {
    // TODO: Implement parsing logic
    // Placeholder implementation
    final pageHeight = XmlHelper.getElementTextAsDouble(
        element.findElements('page-height').firstOrNull);
    final pageWidth = XmlHelper.getElementTextAsDouble(
        element.findElements('page-width').firstOrNull);
    final margins = <PageMargins>[];
    for (final marginElement in element.findElements('page-margins')) {
      final type = marginElement.getAttribute('type');
      final left = XmlHelper.getElementTextAsDouble(
          marginElement.findElements('left-margin').firstOrNull);
      final right = XmlHelper.getElementTextAsDouble(
          marginElement.findElements('right-margin').firstOrNull);
      final top = XmlHelper.getElementTextAsDouble(
          marginElement.findElements('top-margin').firstOrNull);
      final bottom = XmlHelper.getElementTextAsDouble(
          marginElement.findElements('bottom-margin').firstOrNull);
      margins.add(PageMargins(
        type: type,
        leftMargin: left,
        rightMargin: right,
        topMargin: top,
        bottomMargin: bottom,
      ));
    }
    return PageLayout(
      pageHeight: pageHeight,
      pageWidth: pageWidth,
      pageMargins: margins,
    );
  }
}

/// Parser for <scaling> elements.
class ScalingParser {
  /// Parses an [XmlElement] into a [Scaling] object.
  Scaling parse(XmlElement element) {
    // TODO: Implement parsing logic if needed separately, or integrate into ScoreParser/DefaultsParser
    // Placeholder implementation
    final millimeters = XmlHelper.getElementTextAsDouble(
        element.findElements('millimeters').firstOrNull);
    final tenths = XmlHelper.getElementTextAsDouble(
        element.findElements('tenths').firstOrNull);

    if (millimeters == null || tenths == null) {
      // Or throw an exception, depending on how strict parsing should be
      // For now, returning a default or handling nullability in the Scaling model might be options
      // However, the model expects non-null values.
      throw Exception(
          '<scaling> element must contain <millimeters> and <tenths>');
    }
    return Scaling(millimeters: millimeters, tenths: tenths);
  }
}

File: csharp/MusicXMLParser/Parser/PageLayoutParser.cs
// Assuming necessary using statements for MusicXML models and helpers
using System.Xml.Linq;
using System.Collections.Generic;
using System.Linq;
using MusicXMLParser.Models; // For PageLayout, PageMargins
using MusicXMLParser.Utils; // For XmlHelper

namespace MusicXMLParser.Parser
{
    /// <summary>
    /// Parser for <page-layout> elements.
    /// </summary>
    public class PageLayoutParser
    {
        public PageLayout Parse(XElement element)
        {
            var pageHeight = XmlHelper.GetElementTextAsDouble(element.Elements("page-height").FirstOrDefault());
            var pageWidth = XmlHelper.GetElementTextAsDouble(element.Elements("page-width").FirstOrDefault());
            var margins = new List<PageMargins>();

            foreach (var marginElement in element.Elements("page-margins"))
            {
                var type = marginElement.Attribute("type")?.Value;
                var left = XmlHelper.GetElementTextAsDouble(marginElement.Elements("left-margin").FirstOrDefault());
                var right = XmlHelper.GetElementTextAsDouble(marginElement.Elements("right-margin").FirstOrDefault());
                var top = XmlHelper.GetElementTextAsDouble(marginElement.Elements("top-margin").FirstOrDefault());
                var bottom = XmlHelper.GetElementTextAsDouble(marginElement.Elements("bottom-margin").FirstOrDefault());

                // Using object initializer syntax as PageMargins relies on public setters
                margins.Add(new PageMargins
                {
                    Type = type,
                    LeftMargin = left,
                    RightMargin = right,
                    TopMargin = top,
                    BottomMargin = bottom
                });
            }

            // Using object initializer syntax as PageLayout relies on public setters
            return new PageLayout
            {
                PageHeight = pageHeight,
                PageWidth = pageWidth,
                // Ensure PageMargins is always initialized, even if empty, to avoid null reference on the list itself.
                PageMargins = margins.Any() ? margins : new List<PageMargins>()
            };
        }
    }
}
